### DISCUSSION TITLE
Reporting parsing erros

### DISCUSSION THREAD

**Discussion by dubik** [petitparser/dart-petitparser#185] [Q&A] [answered]
I saw previous question about this topic https://github.com/petitparser/dart-petitparser/issues/98 but unfortunately it doesn't seem to work, or at least I don't understand what I'm doing wrong.
I'm trying to parse protobuf like grammar:

```dart
class ProtobufGrammar extends GrammarDefinition {
  @override
  Parser start() => ref0(file).end();

  Parser file() => ref0(definition).star();

  Parser definition() => ref0(message);

  Parser message() =>
      string('message').trim() &
      ref0(identifier).trim() &
      char('{').trim() &
      char('}').trim();

  Parser identifier() => (letter() & word().star())
      .flatten("identifier is expected");
}

void main() {
  final parser = ProtobufGrammar().build();
  test('message without name', () {
    final input = 'message ';
    final result = parser.parse(input);
    // Would be great to have an error: identifier is expected
    expect(result, isSuccess());
  });

  test('message with name', () {
    final input = 'message Empty';
    final result = parser.parse(input);
    // Would be great to have an error: '{' is expected
    expect(result, isSuccess());
  });

  test('Bad identifier', () {
    final input = 'message f123';
    final result = parser.parse(input);
    expect(result, isSuccess());
  });
}
```
All test cases fail (that's correct, since input is malformed), with the same error message: `<Failure[1:1]: end of input expected>`
After reading #98 I would expected `Bad identifer` to report `identifier is expected` but it's not. I tried to use `.or(failure("identifier is expected")` but it didn't do anything.

I think it goes into: 
```dart
  Parser message() =>
      string('message').trim() &
```
and since it can't satisfy `ref0(identifier).trim()` it bails out, and tries another parser. But since there is no alternative with `message` (there is single parser anyway), I would like it to fail with something else than generic `end of input expected`. Is there a way, thanks?

**Comment by dubik** [comment]
It seems to work the way I expect if I omit`star()` or replace it with `plus()` from `Parser file() => ref0(definition).star();` e.g. I'm expecting one message definition. It correctly shows errors. But I may have empty file, or I may some other rules like comments so _zero or more_ have to be there (somewhere).

**Comment by renggli** [comment, marked as answer]
Thank you for the one-click reproducible example. This is really appreciated!

As you observed, the issue is that the `star()` operator is happy to end before a failing definition, which then causes the end-of-input parser to fail with a not so helpful error message.

You can avoid the problem after the repeating `star()` operator by checking that you either

- reached the end-of-input (which is good), OR
- you retry the definition to pass up a nice error message.

```dart
Parser start() => ref0(file);
Parser file() => ref0(definition).star() & (endOfInput().trim() | ref0(definition));
...
```

**Comment by dubik** [comment]
Thanks for a quick reply, seems to work!
